File: src/Must_Be_Flag.cpp
#include <Rcpp.h>
#include "functions.h"
using namespace Rcpp;





void Must_Be_Flag (const std::string& PRED, const SEXP& flag) {

	if (TYPEOF(flag) != LGLSXP) {

		const std::string message =
			"The predicate function " + dquote(PRED) + \
			" produced a non-{True, False, Na} value.\n";

		Function error_callback("error_callback");
		error_callback(message, 1);

	} else {

		const LogicalVector logical_flag = as<LogicalVector>(flag);

		if (logical_flag.size() != 1) {

			std::stringstream msg;
			msg << "The predicate function "
				<< dquote(PRED)
				<< " produced a non-{True, False, Na} value.\n";

			Function error_callback("error_callback");
			error_callback(msg.str(), 1);

		}

	}
}

File: R/must-be-flag.R

Must_Be_Flag <- function (BOOL, PRED) {
	# this macro expands to check if a value is True, False or Na.

	BOOL <- substitute(BOOL)
	PRED <- substitute(PRED)

	bquote( if (!is.logical( .(BOOL) ) || length( .(BOOL) ) != 1) {

		message <-
			"The predicate function " %+% ddquote( .(PRED) ) %+%
			" produced a non-{True, False, Na} value." %+%
			summate( .(BOOL) )

		throw_exception $ error(sys.call(), message)

	} )
}
